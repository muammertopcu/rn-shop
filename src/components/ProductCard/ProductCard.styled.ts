import styled from 'styled-components/native';
import {Text, View, Image, TouchableOpacity} from 'react-native';
import {moderateScale} from 'react-native-size-matters';

export const Container = styled(TouchableOpacity)`
  gap: ${moderateScale(10)}px;
  flex: 1;
`;

export const ProductImage = styled(Image)`
  aspect-ratio: 1;
  border-radius: ${moderateScale(25)}px;
  background-color: white;
`;

export const ProductInfo = styled(View)`
  gap: ${moderateScale(2)}px;
  align-items: center;
`;

export const ProductName = styled(Text).attrs({numberOfLines: 1})`
  color: #000;
  font-weight: bold;
  font-size: ${moderateScale(14)}px;
  padding: 0 ${moderateScale(10)}px;
`;

export const Category = styled(Text)`
  color: #666666;
  text-transform: capitalize;
`;

export const Price = styled(Text)`
  color: #000;
  font-weight: bold;
  font-size: ${moderateScale(16)}px;
`;
