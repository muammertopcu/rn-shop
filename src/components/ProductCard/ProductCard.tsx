import React from 'react';
import {
  Category,
  Container,
  Price,
  ProductImage,
  ProductInfo,
  ProductName,
} from './ProductCard.styled.ts';
import type {ProductCardProps} from './ProductCard.types.ts';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';

const ProductCard = ({product}: ProductCardProps) => {
  const navigation = useNavigation<NativeStackNavigationProp<HomeStackParamList>>();

  return (
    <Container
      testID={'product-card-container'}
      onPress={() => navigation.navigate('ProductDetail', {productId: product.id.toString()})}>
      <ProductImage source={{uri: product.thumbnail}} />

      <ProductInfo>
        <ProductName>{product.title}</ProductName>
        <Category>{product.category}</Category>
        <Price>${product.price}</Price>
      </ProductInfo>
    </Container>
  );
};

export default ProductCard;
