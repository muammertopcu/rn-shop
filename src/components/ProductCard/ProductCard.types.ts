import type {Product} from '@types';

export interface ProductCardProps {
  product: Product;
}
