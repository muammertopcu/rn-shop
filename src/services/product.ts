import api from './api.ts';
import type {
  PaginationParams,
  PaginationResponse,
  Product,
  ProductResponse,
  ProductSearchParams,
} from '@types';

const productApiSlice = api.injectEndpoints({
  endpoints: build => ({
    getProduct: build.query<
      PaginationResponse<ProductResponse>,
      PaginationParams<ProductSearchParams>
    >({
      query: params => ({url: 'products/search', params}),
      providesTags: result =>
        result
          ? [
              ...result.products.map(({id}) => ({type: 'Product' as const, id})),
              {type: 'Product', id: 'LIST'},
            ]
          : [{type: 'Product', id: 'LIST'}],
      serializeQueryArgs: ({queryArgs}) => {
        const newQueryArgs = {...queryArgs};

        if (typeof newQueryArgs.skip === 'number') {
          delete newQueryArgs.skip;
        }

        return newQueryArgs;
      },
      merge: (currentCacheData, responseData) => {
        if (!currentCacheData) {
          return responseData;
        }

        const uniqueProducts = responseData.products.filter(
          product => !currentCacheData.products.some(({id}) => id === product.id),
        );

        return {
          ...responseData,
          products: [...currentCacheData.products, ...uniqueProducts],
        };
      },
    }),
    getProductDetail: build.query<Product, string>({
      query: id => ({url: `products/${id}`}),
      providesTags: (_result, _error, id) => [{type: 'Product', id}],
    }),
  }),
  overrideExisting: true,
});

export const {useLazyGetProductQuery, useGetProductDetailQuery} = productApiSlice;
