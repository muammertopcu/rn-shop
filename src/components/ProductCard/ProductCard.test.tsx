import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import ProductCard from './ProductCard.tsx';

import product from '__mocks__/product.json';

describe('<ProductCard />', () => {
  it('should render correctly', () => {
    const {getByTestId, toJSON} = render(<ProductCard product={product} />);

    expect(getByTestId('product-card-container')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render the product info', () => {
    const {getByText} = render(<ProductCard product={product} />);

    expect(getByText(product.title)).toBeTruthy();
    expect(getByText(product.category)).toBeTruthy();
    expect(getByText(`$${product.price}`)).toBeTruthy();
  });

  it('should navigate to the product detail screen', () => {
    const mockNavigate = jest.fn();

    jest
      .spyOn(require('@react-navigation/native'), 'useNavigation')
      .mockReturnValue({navigate: mockNavigate});

    const {getByTestId} = render(<ProductCard product={product} />);

    fireEvent(getByTestId('product-card-container'), 'onPress');

    expect(mockNavigate).toHaveBeenCalledWith('ProductDetail', {productId: product.id.toString()});
  });
});
