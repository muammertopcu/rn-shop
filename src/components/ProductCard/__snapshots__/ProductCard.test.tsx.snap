// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
      "gap": 15.714285714285714,
      "opacity": 1,
    }
  }
  testID="product-card-container"
>
  <Image
    source={
      {
        "uri": "https://via.placeholder.com/150",
      }
    }
    style={
      {
        "aspectRatio": 1,
        "backgroundColor": "white",
        "borderBottomLeftRadius": 39.285714285714285,
        "borderBottomRightRadius": 39.285714285714285,
        "borderTopLeftRadius": 39.285714285714285,
        "borderTopRightRadius": 39.285714285714285,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "gap": 3.142857142857143,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#000",
          "fontSize": 22,
          "fontWeight": "bold",
          "paddingBottom": 0,
          "paddingLeft": 15.714285714285714,
          "paddingRight": 15.714285714285714,
          "paddingTop": 0,
        }
      }
    >
      Essence Mascara Lash Princess
    </Text>
    <Text
      style={
        {
          "color": "#666666",
          "textTransform": "capitalize",
        }
      }
    >
      beauty
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 25.142857142857142,
          "fontWeight": "bold",
        }
      }
    >
      $
      9.99
    </Text>
  </View>
</View>
`;
